File: leetcode/company-wise/meta/0528_random_pick_with_weight.py
# LC 528. Random Pick with Weight | Medium | Meta
import bisect, random

class Solution:
    def __init__(self, w):
        self.prefix = []
        total = 0
        for weight in w: total += weight; self.prefix.append(total)
        self.total = total

    def pickIndex(self):
        target = random.random() * self.total
        return bisect.bisect_right(self.prefix, target)

File: leetcode/company-wise/meta/0528_random_pick_with_weight.cpp
// LC 528. Random Pick with Weight | Medium | Meta
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
    vector<int> prefix;
public:
    Solution(vector<int>& w) : prefix(w) {
        for (int i = 1; i < w.size(); i++) prefix[i] += prefix[i-1];
    }
    int pickIndex() {
        int target = rand() % prefix.back();
        return upper_bound(prefix.begin(), prefix.end(), target) - prefix.begin();
    }
};
